metricflow/plan_conversion/to_sql_plan/dataflow_to_subquery.py

```diff
@@ -163,6 +167,30 @@ def _next_unique_cte_alias(self) -> str:
         """Return the next unique CTE alias to use in generating queries."""
         return SequentialIdGenerator.create_next_id(StaticIdPrefix.CTE).str_value
 
+    def get_output_data_set(self, node: DataflowPlanNode) -> SqlDataSet:
+        """Cached since this will be called repeatedly during the computation of multiple metrics.
+
+        # TODO: The cache needs to be pruned, but has not yet been an issue.
+        """
+        if node not in self._node_to_output_data_set:
+            self._node_to_output_data_set[node] = node.accept(self)
+
+        return self._node_to_output_data_set[node]
```

Comment:
If the cache is used in the conversion to SQL, this will will need to return a `.copy()` of the SQL node to avoid duplication of nodes in the generated plan.

Reply:
When would you expect duplicate nodes to be generated? Also, do you mean dataflow plan nodes or SQL nodes?

Reply:
That should only happen if we have the exact same dataflow plan node, right? Same node ID and everything? I'm just worried if we return a copy, we won't actually hit the cache as often (but maybe I'm wrong about that).

Reply:
> That should only happen if we have the exact same dataflow plan node, right? Same node ID and everything?
Yes.

>  I'm just worried if we return a copy, we won't actually hit the cache as often (but maybe I'm wrong about that).

So hitting the cache will be determined by `node`, which shouldn't change. To clarify, the copy refers to copying the SQL node in `SqlDataSet`.

Reply:
Ok makes sense! Updated. With the snapshot changes in the latest commit now I see the problem I had overlooked in the prior snapshots.
Should be ready for review again!